spotlight-client/src/datastores/Metric.ts

```diff
@@ -0,0 +1,321 @@
+        (record) => record.locality === this.localityId
+      );
+    }
+
+    if (this.demographicView) {
+      recordsToReturn = recordsToReturn.filter(
+        recordIsTotalByDimension(this.demographicView)
+      );
+    }
+
+    return recordsToReturn;
+  }
+}
+
+export default Metric;
+
+/**
+ * Returns a filter predicate for the specified demographic view
+ * that will exclude totals and breakdowns for all other views
+ */
+function recordIsTotalByDimension(
+  demographicView: DemographicView
+): (record: RawRecord) => boolean {
+  const keysEnum = { ...DIMENSION_DATA_KEYS };
+
+  if (demographicView !== "total") {
+    delete keysEnum[demographicView];
+  }
+
+  const otherDataKeys = Object.values(keysEnum);
+
+  return (record) => {
+    let match = true;
+    if (demographicView !== "total") {
+      // filter out totals
+      match =
+        match && record[DIMENSION_DATA_KEYS[demographicView]] !== TOTAL_KEY;
+    }
+
+    // filter out subset permutations
+    match = match && otherDataKeys.every((key) => record[key] === TOTAL_KEY);
+
+    return match;
+  };
+}
+
+export type MetricsMapping = {
+  [MetricTypes.SentencePopulationCurrent]?: PopulationSnapshot;
+  [MetricTypes.SentenceTypesCurrent]?: SentenceTypesCurrent;
+};
+
+function extractBaseFields(record: ValuesType<DataFile>): BaseMetricFields {
+  return {
+    stateCode: record.state_code,
+  };
+}
+
+function extractDemographicFields(
+  record: ValuesType<DataFile>
+): DemographicFields {
+  return {
+    // we are trusting the API to return valid strings here, not validating them
+    raceOrEthnicity: record.race_or_ethnicity as RaceIdentifier,
+    gender: record.gender as GenderIdentifier,
+    ageBucket: record.age_bucket as AgeIdentifier,
+  };
+}
+
+function transformSentencePopulationCurrent(
+  rawRecords: DataFile
+): PopulationBreakdownByLocationRecord[] {
+  return rawRecords.map((record) => {
+    return {
+      locality: record.district,
+      population: Number(record.total_population_count),
+      ...extractBaseFields(record),
+      ...extractDemographicFields(record),
+    };
+  });
+}
+
+function transformSentenceTypes(
+  rawRecords: DataFile
+): SentenceTypeByLocationRecord[] {
+  return rawRecords.map((record) => {
+    return {
+      dualSentenceCount: Number(record.dual_sentence_count),
+      incarcerationCount: Number(record.incarceration_count),
+      locality: record.district,
+      probationCount: Number(record.probation_count),
+      ...extractBaseFields(record),
+      ...extractDemographicFields(record),
+    };
+  });
+}
+
+export function createMetric(opts: {
+  metricType: MetricTypes.SentencePopulationCurrent;
+  initOptions: MetricFactoryOptionsBase;
+}): PopulationSnapshot;
+export function createMetric(opts: {
+  metricType: MetricTypes.SentenceTypesCurrent;
+  initOptions: MetricFactoryOptionsBase;
+}): SentenceTypesCurrent;
+// TODO: like 20 more overloads for all the different metric types -_-
```

Comment:
the further I get into implementing this feature the less I am bothered by this, honestly. It seems to be a pretty common pattern in Typescript for "narrowing" broad types to more specific instances. It does actually seem like the most straightforward way of splitting the difference between high-dimensional types at the core and specific types at the edges. At the very least it is _painfully_ obvious what this code is doing, so while it may not be super fun to write the first time it may not be all that bad in terms of ongoing maintenance